server/filestore.go

```diff
@@ -770,25 +773,42 @@ func (mb *msgBlock) selectNextFirst() {
 	}
 }
 
-func (fs *fileStore) deleteMsgFromBlock(mb *msgBlock, seq uint64, sm *fileStoredMsg, secure bool) {
+func (fs *fileStore) deleteMsgFromBlock(mb *msgBlock, seq uint64, sm *fileStoredMsg, secure bool) error {
 	// Update global accounting.
 	msz := fileStoreMsgSize(sm.subj, sm.hdr, sm.msg)
 
 	fs.mu.Lock()
 	mb.mu.Lock()
 
 	// Make sure the seq still exists.
+	if mb.cache == nil {
+		// Rare condition but can happen. We will load with lock held at this point.
+		buf, err := ioutil.ReadFile(mb.mfn)
+		if err == nil {
+			err = mb.indexCacheBuf(buf)
+		}
+		if err != nil {
+			mb.mu.Unlock()
+			fs.mu.Unlock()
+			return err
+		}
+		if len(buf) > 0 {
+			mb.cloads++
+			mb.startCacheExpireTimer()
+		}
+	}
```

Comment:
isn't this done by func (mb *msgBlock) loadMsgs() error { as well? call that instead?

Reply:
It wouldn't hold all the locks either.

Reply:
loadMsgs does but does its own lock handling which is correct for most cases. In this case we need to hold the lock to ensure either we error or we have the cache filled. I could make it trip but only artificially.